### ISSUE TITLE
Wrong FLOPS calculation for inferencing different image size

### ISSUE THREAD

**Issue by AlanLi1997** [ultralytics/ultralytics#9350]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

I found an error when I tried to verify model accuracy with different image resolutions. FLOPs seems to be fixed to a constant, however, in fact it should be positively related to the validation image size. The relevant script I found should be _torch_utils.py_. After I fixed, the FLOPs calculation became correct.

### Minimal Reproducible Example

# an example, image size =1280
def get_flops(model, imgsz=640):
    """Return a YOLO model's FLOPs."""
    try:
        model = de_parallel(model)
        p = next(model.parameters())
        stride = max(int(model.stride.max()), 32) if hasattr(model, 'stride') else 32  # max stride
        im = torch.empty((1, p.shape[1], stride, stride), device=p.device)  # input image in BCHW format
        flops = thop.profile(deepcopy(model), inputs=[im], verbose=False)[0] / 1E9 * 2 if thop else 0  # stride GFLOPs
        imgsz = imgsz if isinstance(imgsz, list) else [imgsz, imgsz]  # expand if int/float
        #flops * imgsz[0] / stride * imgsz[1] / stride  # 640x640 GFLOPs
        return flops * 1280 / stride * 1280 / stride  # 1280x1280 GFLOPs
    except Exception:
        return 0

def get_flops_with_torch_profiler(model, imgsz=640):
    """Compute model FLOPs (thop alternative)."""
    if TORCH_2_0:
        model = de_parallel(model)
        p = next(model.parameters())
        stride = (max(int(model.stride.max()), 32) if hasattr(model, 'stride') else 32) * 2  # max stride
        im = torch.zeros((1, p.shape[1], stride, stride), device=p.device)  # input image in BCHW format
        with torch.profiler.profile(with_flops=True) as prof:
            model(im)
        flops = sum(x.flops for x in prof.key_averages()) / 1E9
        imgsz = imgsz if isinstance(imgsz, list) else [imgsz, imgsz]  # expand if int/float
        flops = flops * imgsz[0] / stride * imgsz[1] / stride  # 640x640 GFLOPs
        return flops
    return 0

### Additional

If this error has been fixed in a new version, please just close this issue.

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@AlanLi1997 thank you for bringing this to our attention! 🛠️ It looks like you've found an inconsistency in the FLOPs calculation with varying image sizes. Your observation is spot-on; FLOPs should indeed scale with the dimensions of the input images.

To properly handle different input image sizes, the FLOPs calculation should reflect the actual computation executed by the model. Your adjustment to the `get_flops` function seems appropriate for fixing the issue for a specific input size (1280x1280 in your example). 

For a more general solution, ensuring the FLOPs are calculated dynamically based on `imgsz` would ensure accuracy across various resolutions. Something like this:

```python
# Corrected dynamic image size handling for FLOPs calculation
return flops * imgsz[0] / stride * imgsz[1] / stride  # General GFLOPs calculation
```
This change ensures the FLOPs calculation accurately scales with the input image size provided in `imgsz`.

Your willingness to contribute with a PR is much appreciated! 🙏 If you could include a dynamic handling of different image resolutions in your PR, that would be fantastic.

Just as a side note, make sure your PR includes tests or example cases demonstrating the fix for different scenarios of image sizes. This way, we ensure comprehensive coverage and maintain the robustness of the model.

Thanks again for your contribution, looking forward to your PR! If there's anything more you'd like to discuss or need further clarification on, feel free to ask.

**Comment by glenn-jocher** [comment]
@AlanLi1997 all done! THank you for the PR :)

**Comment by glenn-jocher** [comment]
@AlanLi1997 yes you are correct, FLOPs here is fixed to image size. I'll take a look at how to improve this.
